text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+     */
+    metricInvocationResponseCode(responseCode: InvocationHttpResponseCode, props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for disk utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricDiskUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for CPU utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricCPUUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for memory utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricMemoryUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for GPU utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricGPUUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for GPU memory utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricGPUMemoryUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Enable autoscaling for SageMaker Endpoint production variant
+     *
+     * @param scalingProps EnableScalingProps
+     */
+    autoScaleInstanceCount(scalingProps: appscaling.EnableScalingProps): ScalableInstanceCount;
+  }
+
+  class EndpointProductionVariant implements IEndpointProductionVariant { ... }
+  ```
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
+
+1. Theoretically, the `ContainerImage` code (referenced [above](#container-image)) from the
+   `@aws-cdk/ecs` and `@aws-cdk/sagemaker` modules could be unified assuming it would be sufficient
+   for both use-cases to `bind` using an `IGrantable` (adjusting ECS's `TaskDefinition`
+   accordingly). However, it's unclear within which module such a unified API should reside as
+   support for private repositories makes it a bad fit for `@aws-cdk/ecr` and `@aws-cdk/ecr-assets`,
+   and it would be unintuitive for `@aws-cdk/sagemaker` to declare a dependency on `@aws-cdk/ecs`.
+
+   Package concerns aside, historically, there was a period during which SageMaker only supported
+   ECR as an image source while ECS was capable of sourcing images from either ECR or a
+   customer-owned private repository. Given the fact that these two products' supported images
+   sources may yet again diverge in the future, maybe it would be best to keep their
+   `ContainerImage` APIs separate within their respective modules.
+1. In the [earliest PR][earliest-pr] attempting to add SageMaker L2 constructs to the CDK, the
+   author did not create an `EndpointConfig` construct, instead hiding the resource's creation
+   behind `Endpoint` (to which production variants could be added). Although a simplifier, this
+   prevents customers from reusing configuration across endpoints. For this reason, an explicit
+   L2 construct for endpoint configuration was incorporated into this RFC.
```

Comment:
Thanks for getting into this. What would be a typical use case to use the same config for multiple endpoints?

Reply:
Responded on a different comment but this makes sense, just please upstream it to the EndpointConfig readme section (just a simple "By using the `EndpointConfig` construct, you can define one endpoint and reuse it on multiple `Endpoint` constructs")

Reply:
In rev2, I now mention reuse of `EndpointConfig` across `Endpoint` resources in the README.